File: array/3/3.longest-substring-without-repeating-characters.c
// @leet start
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
int lengthOfLongestSubstring(char *s) {
  int n = strlen(s);
  if (n == 0) {
    return 0;
  }

  int maxLength = 0;
  for (int i = 0; i < n; i++) {
    bool visited[256] = {false}; // 标记字符是否出现过
    int length = 0;
    for (int j = i; j < n; j++) {
      if (visited[(unsigned char)s[j]]) {
        break;
      }
      visited[(unsigned char)s[j]] = true;
      length++;
    }
    if (length > maxLength) {
      maxLength = length;
    }
  }
  return maxLength;
}
// @leet end

File: array/3/3.longest-substring-without-repeating-characters.py
# @leet start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == "":
            return 0
        stack = []
        maxLength = 0
        for i in range(len(s)):
            #     if s[i] not in stack:
            #         stack.append(s[i])
            #         if len(stack) > maxLength:
            #             maxLength = len(stack)
            #     else:
            #         # if len(stack) > maxLength:
            #         #     maxLength = len(stack)
            #         stack.clear()
            #         stack.append(s[i])
            # return maxLength

            for j in range(i+1, len(s)):
                if s[j] not in stack:
                    stack.append(s[j])
                    # print(stack)
                else:
                    i = j
                    if len(stack) > maxLength:
                        maxLength = len(stack)
                    stack.clear()
                    break
            if len(stack) > maxLength:
                maxLength = len(stack)
            stack.clear()
        return maxLength

# result=Solution().lengthOfLongestSubstring("")
# print(result)
# @leet end
